Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h

```diff
@@ -0,0 +1,438 @@
+            return L;
+        // variants can be used by any of the underlying types
+        // thus recursively use this function
+        else if constexpr (is_variant<T>::value)
+        {
+            int iMatch = TypeMatchVariant<T>(L, index);
+            return PopUnsafeVariant<T>(L, index, iMatch);
+        }
+
+        // Catch all for class pointer types, assume all classes are valid script entities
+        // and can be fetched from a userdata
+        else if constexpr (std::is_pointer_v<T> && std::is_class_v<std::remove_pointer_t<T>>)
+        {
+            bool  isLightUserData = lua_type(L, index) == LUA_TLIGHTUSERDATA;
+            void* pValue = lua::PopTrivial<void*>(L, index);
+            using class_t = std::remove_pointer_t<T>;
+            auto result =
+                isLightUserData ? UserDataCast<class_t>((class_t*)0, pValue, L) : UserDataCast<class_t>((class_t*)0, *reinterpret_cast<void**>(pValue), L);
+            if (result == nullptr)
+            {
+                SString strReceived = isLightUserData ? GetUserDataClassName(pValue, L) : GetUserDataClassName(*(void**)pValue, L);
+                SString strExpected = GetClassTypeName((T)0);
+                SString strMessage("Expected %s at argument %d, got %s", strExpected.c_str(), index, strReceived.c_str());
+                strError = strMessage;
+                return nullptr;
+            }
+            return static_cast<T>(result);
+        }
+    }
+};
+
+template <bool, auto*>
+struct CLuaFunctionParser
+{
+};
+
+template <bool ErrorOnFailure, typename Ret, typename... Args, auto (*Func)(Args...)->Ret>
+struct CLuaFunctionParser<ErrorOnFailure, Func> : CLuaFunctionParserBase
+{
+    template <typename... Params>
+    inline auto Call(lua_State* L, Params&&... ps)
+    {
+        if (strError.length() != 0)
+        {
+            return -1;
+        }
+        if constexpr (sizeof...(Params) == sizeof...(Args))
+        {
+            if constexpr (std::is_same_v<Ret, void>)
+            {
+                Func(std::forward<Params>(ps)...);
+                return 0;
+            }
+            else
+            {
+                return lua::Push(L, Func(std::forward<Params>(ps)...));
+            }
+        }
+        else
+        {
+            return Call(L, ps..., Pop<typename nth_element_impl<sizeof...(Params), Args...>::type>(L, iIndex));
+        }
+    }
+
+    inline int operator()(lua_State* L, CScriptDebugging* pScriptDebugging)
+    {
+        int iResult = 0;
+        try
+        {
+            iResult = Call(L);
+        }
+        catch (std::invalid_argument& e)
+        {
+            strError = e.what();
+        }
+        if (strError.length() != 0)
+        {
+            if constexpr (ErrorOnFailure)
+            {
+                luaL_error(L, strError.c_str());
+            }
+            else
+            {
+                pScriptDebugging->LogCustom(L, strError.c_str());
+                lua_pushboolean(L, false);
```

Comment:
If I understood it correctly, in "warning-mode" this always returns _false_.   
As mentioned in https://github.com/multitheftauto/mtasa-blue/issues/821#issue-411212784 we probably need yet another case (in which _nil_ is returned) though.

Reply:
Fixed (additional template parameter for ArgumentParserWarn added). You can now either do 

- `ArgumentParserWarn<nullptr, func>` to return nil on failure
- `ArgumentParserWarn<false, func>` to return false on failure

I don't think we need more, do we?